mesonbuild/modules/gobuild.py

```diff
@@ -0,0 +1,252 @@
+        "gobuild.project",
+        str,
+        varargs=(
+            str,
+            mesonlib.File,
+            build.CustomTarget,
+            build.CustomTargetIndex,
+            build.GeneratedList,
+        ),
+    )
+    @typed_kwargs(
+        "gobuild.project",
+        CT_INPUT_KW.evolve(name="sources"),
+        KwargInfo(
+            "include_directories",
+            ContainerTypeInfo(list, (str, build.IncludeDirs)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "link_with",
+            ContainerTypeInfo(list, (build.SharedLibrary, build.StaticLibrary)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "tags",
+            ContainerTypeInfo(list, (str)),
+            default=[],
+            listify=True,
+        ),
+        KwargInfo(
+            "native",
+            bool,
+            default=False,
+        ),
+        KwargInfo(
+            "gomod_tidy",
+            bool,
+            default=True,
+        ),
+    )
+    def project(
+        self,
+        state: ModuleState,
+        args: T.Tuple[
+            str, T.List[T.Union[mesonlib.FileOrString, build.GeneratedTypes]]
+        ],
+        kwargs: ProjectKwargs,
+    ) -> ModuleReturnValue:
+        is_cross_build = state.environment.is_cross_build(
+            mesonlib.MachineChoice.BUILD
+            if kwargs["native"]
+            else mesonlib.MachineChoice.HOST
+        )
+
+        cross_compile_tools = self.detect_tools(state, is_cross_build)
+
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs["sources"]))
+        )
+        go_sum_target = build.CustomTarget(
+            f"{proj_name}: go mod tidy",
+            state.subdir,
+            state.subproject,
+            state.environment,
+            [
+                self.go_exe,
+                "-C",
+                "@CURRENT_SOURCE_DIR@",
+                "mod",
+                "tidy" if kwargs["gomod_tidy"] else "verify",
+            ],
+            self.interpreter.source_strings_to_files(["go.mod"]),
+            [f"go-mod-tidy-{proj_name}.log"],
+            capture=True,
+            build_always_stale=True,
+        )
+
+        build_dir_abs = Path(state.environment.get_build_dir())
+        target_dir_abs = build_dir_abs / state.backend.get_target_dir(go_sum_target)
+
+        link_dirs = {
+            "-L" + (build_dir_abs / state.backend.get_target_dir(lib)).as_posix()
+            for lib in kwargs["link_with"]
+        }
+        link_targets = {"-l" + lib.name for lib in kwargs["link_with"]}
+
+        if not is_cross_build:
+            extra_env = {}
+        else:
+            assert "c" in cross_compile_tools
+            assert "cpp" in cross_compile_tools
+
+            extra_env = {
+                "GOOS": "windows",
+                "GOARCH": "amd64",
```

Comment:
Help wanted to retrieve the `host_machine.system()` from DSL. Also, map `x86_64` to `amd64` because CGO only honors `amd64`.

Reply:
`state.environment.machines` is the backing store for the various `*_machine` DSL constructs. you can access it by indexing with `MachineChoice` or by `.host` or `.build`. Those are instances of the `MachineInfo` class, if you want to lookup the specific methods.